src/github.com/matrix-org/dendrite/mediaapi/thumbnailer/thumbnailer.go

```diff
@@ -0,0 +1,416 @@
+		return false, activeThumbnailGenerationResult.Err
+	}
+
+	// No active thumbnail generation so create one
+	activeThumbnailGeneration.PathToResult[string(dst)] = &types.ThumbnailGenerationResult{
+		Cond: &sync.Cond{L: activeThumbnailGeneration},
+	}
+
+	return true, nil
+}
+
+// broadcastGeneration broadcasts that thumbnail generation completed and the error to all waiting goroutines
+// Note: This should only be called by the owner of the activeThumbnailGenerationResult
+func broadcastGeneration(dst types.Path, activeThumbnailGeneration *types.ActiveThumbnailGeneration, config types.ThumbnailSize, errorReturn error, logger *log.Entry) {
+	activeThumbnailGeneration.Lock()
+	defer activeThumbnailGeneration.Unlock()
+	if activeThumbnailGenerationResult, ok := activeThumbnailGeneration.PathToResult[string(dst)]; ok {
+		logger.WithFields(log.Fields{
+			"Width":        config.Width,
+			"Height":       config.Height,
+			"ResizeMethod": config.ResizeMethod,
+		}).Info("Signalling other goroutines waiting for this goroutine to generate the thumbnail.")
+		// Note: retErr is a named return value error that is signalled from here to waiting goroutines
+		activeThumbnailGenerationResult.Err = errorReturn
+		activeThumbnailGenerationResult.Cond.Broadcast()
+	}
+	delete(activeThumbnailGeneration.PathToResult, string(dst))
+}
+
+// resize scales an image to fit within the provided width and height
+// If the source aspect ratio is different to the target dimensions, one edge will be smaller than requested
+// If crop is set to true, the image will be scaled to fill the width and height with any excess being cropped off
+func resize(dst types.Path, buffer []byte, w, h int, crop bool, logger *log.Entry) (int, int, error) {
+	inImage := bimg.NewImage(buffer)
+
+	inSize, err := inImage.Size()
+	if err != nil {
+		return -1, -1, err
+	}
+
+	options := bimg.Options{
+		Type:    bimg.JPEG,
+		Quality: 85,
+	}
+	if crop {
+		options.Width = w
+		options.Height = h
+		options.Crop = true
+	} else {
+		inAR := float64(inSize.Width) / float64(inSize.Height)
+		outAR := float64(w) / float64(h)
+
+		if inAR > outAR {
+			// input has wider AR than requested output so use requested width and calculate height to match input AR
+			options.Width = w
+			options.Height = int(float64(w) / inAR)
+		} else {
+			// input has narrower AR than requested output so use requested height and calculate width to match input AR
+			options.Width = int(float64(h) * inAR)
+			options.Height = h
+		}
+	}
+
+	newImage, err := inImage.Process(options)
+	if err != nil {
+		return -1, -1, err
+	}
+
+	if err = bimg.Write(string(dst), newImage); err != nil {
+		logger.WithError(err).Error("Failed to resize image")
+		return -1, -1, err
+	}
+
+	return options.Width, options.Height, nil
+}
+
+// init with worst values
+func newThumbnailMetrics() thumbnailMetrics {
+	return thumbnailMetrics{
+		isSmaller:      1,
+		aspect:         float64(16384 * 16384),
```

Comment:
I was actually looking at `MaxFloat64` and previously I was doing some `b.aspect - a.aspect` and was concerned that the precision with such a large number would be such that the subtraction would be inconsequential.

As such, I took something more reasonably within the realm of an acceptable max image resolution. However, now I am comparing the numbers directly I don't have the same precision problem.

I knew this would provoke comment but I forgot to comment it to explain. How do you want me to proceed - comment, `Inf()` or `MaxFloat64`?